File: modules/core/src/data/java/galena/oreganized/data/extensions/OBlockLootExtensions.java
package galena.oreganized.data.extensions;

import com.tterrag.registrate.providers.loot.RegistrateBlockLootTables;

import galena.oreganized.index.sets.StoneSet;
import net.minecraft.core.Holder;
import net.minecraft.world.item.Item;
import net.minecraft.world.level.ItemLike;
import net.minecraft.world.level.block.Block;
import net.minecraft.world.level.block.Blocks;

public class OBlockLootExtensions {

    public static void dropSelf(RegistrateBlockLootTables provider, Holder<? extends Block> block) {
        provider.dropSelf(block.value());
    }

    public static void dropSlab(RegistrateBlockLootTables provider, Holder<? extends Block> slab) {
        provider.add(slab.value(), provider.createSlabItemTable(slab.value()));
    }

    public static void dropOther(RegistrateBlockLootTables provider, Holder<? extends Block> brokenBlock, ItemLike droppedBlock) {
        provider.dropOther(brokenBlock.value(), droppedBlock);
    }

    public static void dropAsSilk(RegistrateBlockLootTables provider, Holder<? extends Block> block) {
        provider.dropWhenSilkTouch(block.value());
    }

    public static void dropWithSilk(RegistrateBlockLootTables provider, Holder<? extends Block> block, Holder<? extends ItemLike> drop) {
        provider.add(block.value(), provider.createSingleItemTableWithSilkTouch(block.value(), drop.value()));
    }

    public static void dropOre(RegistrateBlockLootTables provider, Holder<? extends Block> block, Holder<? extends Item> drop) {
        provider.add(block.value(), provider.createOreDrop(block.value(), drop.value()));
    }

    public static void dropOre(RegistrateBlockLootTables provider, Holder<? extends Block> block, Item drop) {
        provider.add(block.value(), provider.createOreDrop(block.value(), drop));
    }

    public static void dropCauldron(RegistrateBlockLootTables provider, Holder<? extends Block> block) {
        dropOther(provider, block, Blocks.CAULDRON);
    }

    public static void dropNothing(RegistrateBlockLootTables provider, Holder<? extends Block> block) {
        dropOther(provider, block, Blocks.AIR);
    }

    public static void dropPottedPlant(RegistrateBlockLootTables provider, Holder<? extends Block> block) {
        provider.dropPottedContents(block.value());
    }

    public static void dropStoneSet(RegistrateBlockLootTables provider, StoneSet<?, ?, ?, ?> set) {
        dropSelf(provider, set.block());
        dropSlab(provider, set.slab());
        dropSelf(provider, set.stairs());
        dropSelf(provider, set.wall());
    }

}

File: modules/core/src/data/kotlin/galena/oreganized/data/extensions/OBlockLootExtensions.kt
@file:JvmName("OBlockLootExtensions")

package galena.oreganized.data.extensions

import com.tterrag.registrate.providers.loot.RegistrateBlockLootTables
import galena.oreganized.index.sets.StoneSet
import net.minecraft.core.Holder
import net.minecraft.world.item.Item
import net.minecraft.world.level.ItemLike
import net.minecraft.world.level.block.Block
import net.minecraft.world.level.block.Blocks

fun RegistrateBlockLootTables.dropSelf(block: Holder<out Block?>) {
    dropSelf(block.value())
}

fun RegistrateBlockLootTables.dropSlab(slab: Holder<out Block?>) {
    add(slab.value(), createSlabItemTable(slab.value()))
}

fun RegistrateBlockLootTables.dropOther(
    brokenBlock: Holder<out Block?>,
    droppedBlock: ItemLike,
) {
    dropOther(brokenBlock.value(), droppedBlock)
}

fun RegistrateBlockLootTables.dropAsSilk(block: Holder<out Block?>) {
    dropWhenSilkTouch(block.value())
}

fun RegistrateBlockLootTables.dropWithSilk(
    block: Holder<out Block?>,
    drop: Holder<out ItemLike?>,
) {
    add(block.value(), createSingleItemTableWithSilkTouch(block.value(), drop.value()))
}

fun RegistrateBlockLootTables.dropOre(
    block: Holder<out Block?>,
    drop: Holder<out Item?>,
) {
    add(block.value(), createOreDrop(block.value(), drop.value()))
}

fun RegistrateBlockLootTables.dropOre(
    block: Holder<out Block?>,
    drop: Item,
) {
    add(block.value(), createOreDrop(block.value(), drop))
}

fun RegistrateBlockLootTables.dropCauldron(block: Holder<out Block?>) {
    dropOther(block, Blocks.CAULDRON)
}

fun RegistrateBlockLootTables.dropNothing(block: Holder<out Block?>) {
    dropOther(block, Blocks.AIR)
}

fun RegistrateBlockLootTables.dropPottedPlant(block: Holder<out Block?>) {
    dropPottedContents(block.value())
}

fun RegistrateBlockLootTables.dropStoneSet(set: StoneSet<*, *, *, *>) {
    dropSelf(set.block)
    dropSlab(set.slab)
    dropSelf(set.stairs)
    dropSelf(set.wall)
}
